Link error

with `--import-builtin` trying to build this program gets me undefined symbols:

```
Undefined symbols for architecture arm64:
  "_Pointers.945.Pointer.unsafe#925.let", referenced from:
      _Pointers.945.test#942 in Pointers.o
ld: symbol(s) not found for architecture arm64
```

```swift
public type Pointer<Pointee> {

  var base: Builtin.ptr

  memberwise init

  public property unsafe: Pointee {
    yield base as* (remote let Pointee)
  }

}


fun test(p: Pointer<Int>) {
  var x: Int = p.unsafe.copy()
}

public fun main() {}
```

--- a/Sources/Core/TypedProgram.swift
+++ b/Sources/Core/TypedProgram.swift
@@ -122,24 +122,31 @@ public struct TypedProgram: Program {
     relations.monomorphize(generic, for: arguments)
   }
 
-  /// If `t` has a record layout, returns the names and types of its stored properties. Otherwise,
-  /// returns an empty array.
-  public func storage(of t: AnyType) -> [StoredProperty] {
+  /// If `t` has a record layout, returns the names and types of its stored properties, replacing
+  /// generic parameters by their corresponding value in `parameterization`. Otherwise, returns an
+  /// empty array.
+  public func storage(
+    of t: AnyType, parameterizedBy parameterization: GenericArguments = [:]
+  ) -> [StoredProperty] {
     switch t.base {
     case let u as BoundGenericType:
-      return storage(of: u.base)
+      return storage(of: u.base, parameterizedBy: parameterization.appending(u.arguments))
 
     case let u as ProductType:
       return ast[u.decl].members.flatMap { (m) in
         BindingDecl.ID(m).map { (b) in
           ast.names(in: ast[b].pattern).map { (_, name) in
-            (ast[ast[name].decl].baseName, declTypes[ast[name].decl]!)
+            let t = relations.monomorphize(declTypes[ast[name].decl]!, for: parameterization)
+            return (ast[ast[name].decl].baseName, t)
           }
         } ?? []
       }
 
     case let u as TupleType:
-      return u.elements.map({ ($0.label, $0.type) })
+      return u.elements.map({ (l) in
+        let t = relations.monomorphize(l.type, for: parameterization)
+        return (l.label, t)
+      })
 
     default:
       assert(!t.hasRecordLayout)

--- a/Sources/IR/Analysis/Module+AccessReification.swift
+++ b/Sources/IR/Analysis/Module+AccessReification.swift
@@ -149,7 +149,8 @@ extension Module {
     }
 
     let o = RemoteType(k, s.projection.bareType)
-    let reified = makeProject(o, applying: s.variants[k]!, to: arguments, at: s.site)
+    let reified = makeProject(
+      o, applying: s.variants[k]!, parameterizedBy: s.parameterization, to: arguments, at: s.site)
     replace(i, with: reified)
   }
 

--- a/Sources/IR/Analysis/Module+Depolymorphize.swift
+++ b/Sources/IR/Analysis/Module+Depolymorphize.swift
@@ -24,46 +24,55 @@ extension Module {
   private mutating func depolymorphize(_ f: Function.ID, in ir: LoweredProgram) {
     for i in blocks(in: f).map(instructions(in:)).joined() {
       switch self[i] {
-      case let s as CallInstruction:
-        if let new = depolymorphized(s, in: ir) {
-          replace(i, with: new)
-        }
-
-      case let s as ProjectInstruction:
-        if let new = depolymorphized(s, in: ir) {
-          replace(i, with: new)
-        }
-
+      case is CallInstruction:
+        depolymorphize(call: i, in: ir)
+      case is ProjectInstruction:
+        depolymorphize(project: i, in: ir)
       default:
         continue
       }
     }
   }
 
-  /// If `s` is a call to a generic function, returns a new instruction applying a depolymorphized
-  /// version of its callee. Otherwise, returns `nil`.
-  private mutating func depolymorphized(
-    _ s: CallInstruction, in ir: LoweredProgram
-  ) -> CallInstruction? {
+  /// If `i` is a call to a generic function, replaces it by an instruction applying a
+  /// depolymorphized version of its callee. Otherwise, does nothing.
+  ///
+  /// - Requires: `i` identifies a `CallInstruction`
+  private mutating func depolymorphize(
+    call i: InstructionID, in ir: LoweredProgram
+  ) {
+    let s = self[i] as! CallInstruction
     guard
       let callee = s.callee.constant as? FunctionReference,
       !callee.arguments.isEmpty
-    else { return nil }
+    else { return }
 
     // TODO: Use existentialization unless the function is inlinable
 
     let g = monomorphize(callee, in: ir)
     let r = FunctionReference(to: g, usedIn: callee.useScope, in: self)
-    return makeCall(
+    let new = makeCall(
       applying: .constant(r), to: Array(s.arguments), writingResultTo: s.output, at: s.site)
+    replace(i, with: new)
   }
 
-  /// If `s` is the projection through a generic subscript, returns a new instruction applying a
-  /// depolymorphized version of its callee. Otherwise, returns `nil`.
-  private mutating func depolymorphized(
-    _ s: ProjectInstruction, in ir: LoweredProgram
-  ) -> ProjectInstruction? {
-    return nil
+  /// If `i` is the projection through a generic subscript, replaces it by an instruction applying
+  /// a depolymorphized version of its callee. Otherwise, does nothing.
+  ///
+  /// - Requires: `i` identifies a `ProjectInstruction`
+  private mutating func depolymorphize(
+    project i: InstructionID, in ir: LoweredProgram
+  ) {
+    let s = self[i] as! ProjectInstruction
+    guard !s.parameterization.isEmpty else { return }
+
+    // TODO: Use existentialization unless the subscript is inlinable
+
+    let useScope = self[Block.ID(i.function, i.block)].scope
+    let g = monomorphize(s.callee, for: s.parameterization, usedIn: useScope, in: ir)
+    let new = makeProject(
+      s.projection, applying: g, parameterizedBy: [:], to: s.operands, at: s.site)
+    replace(i, with: new)
   }
 
   /// Returns a depolymorphized copy of `base` in which parametric parameters have been notionally
@@ -91,7 +100,17 @@ extension Module {
   private mutating func monomorphize(
     _ r: FunctionReference, in ir: LoweredProgram
   ) -> Function.ID {
-    let result = demandMonomorphizedDeclaration(r)
+    monomorphize(r.function, for: r.arguments, usedIn: r.useScope, in: ir)
+  }
+
+  /// Returns a reference to the monomorphized form of `f` for given `parameterization` in
+  /// `useScope`, reading definitions from `ir`.
+  @discardableResult
+  private mutating func monomorphize(
+    _ f: Function.ID, for parameterization: GenericArguments, usedIn useScope: AnyScopeID,
+    in ir: LoweredProgram
+  ) -> Function.ID {
+    let result = demandMonomorphizedDeclaration(of: f, for: parameterization, in: useScope)
     if self[result].entry != nil {
       return result
     }
@@ -101,17 +120,17 @@ extension Module {
 
     // Iterate over the basic blocks of the source function in a way that guarantees we always
     // visit definitions before their uses.
-    let sourceModule = ir.modules[ir.module(defining: r.function)]!
-    let cfg = sourceModule[r.function].cfg()
-    let sourceBlocks = DominatorTree(function: r.function, cfg: cfg, in: sourceModule).bfs
+    let sourceModule = ir.modules[ir.module(defining: f)]!
+    let cfg = sourceModule[f].cfg()
+    let sourceBlocks = DominatorTree(function: f, cfg: cfg, in: sourceModule).bfs
 
     for b in sourceBlocks {
-      let source = Block.ID(r.function, b)
+      let source = Block.ID(f, b)
 
       // Rewrite the source block in the monomorphized function.
       let scope = sourceModule[source].scope
       let inputs = sourceModule[source].inputs.map { (t) in
-        program.monomorphize(t, applying: r.arguments)
+        program.monomorphize(t, applying: parameterization)
       }
       let target = Block.ID(result, self[result].appendBlock(in: scope, taking: inputs))
       rewrittenBlocks[source] = target
@@ -182,7 +201,7 @@ extension Module {
     /// Rewrites `i`, which is in `r.function`, into `result`, at the end of `b`.
     func rewrite(allocStack i: InstructionID, to b: Block.ID) {
       let s = sourceModule[i] as! AllocStackInstruction
-      let t = program.monomorphize(s.allocatedType, for: r.arguments)
+      let t = program.monomorphize(s.allocatedType, for: parameterization)
       append(makeAllocStack(t, at: s.site), to: b)
     }
 
@@ -218,7 +237,7 @@ extension Module {
     /// Rewrites `i`, which is in `r.function`, into `result`, at the end of `b`.
     func rewrite(callFFI i: InstructionID, to b: Block.ID) {
       let s = sourceModule[i] as! CallFFIInstruction
-      let t = program.monomorphize(s.returnType, applying: r.arguments)
+      let t = program.monomorphize(s.returnType, applying: parameterization)
       let o = s.operands.map(rewritten(_:))
       append(makeCallFFI(returning: t, applying: s.callee, to: o, at: s.site), to: b)
     }
@@ -261,7 +280,7 @@ extension Module {
     /// Rewrites `i`, which is in `r.function`, into `result`, at the end of `b`.
     func rewrite(globalAddr i: InstructionID, to b: Block.ID) {
       let s = sourceModule[i] as! GlobalAddrInstruction
-      let t = program.monomorphize(s.valueType, for: r.arguments)
+      let t = program.monomorphize(s.valueType, for: parameterization)
       append(makeGlobalAddr(of: s.id, in: s.container, typed: t, at: s.site), to: b)
     }
 
@@ -287,7 +306,7 @@ extension Module {
     /// Rewrites `i`, which is in `r.function`, into `result`, at the end of `b`.
     func rewrite(pointerToAddress i: InstructionID, to b: Block.ID) {
       let s = sourceModule[i] as! PointerToAddressInstruction
-      let t = program.monomorphize(^s.target, for: r.arguments)
+      let t = program.monomorphize(^s.target, for: parameterization)
 
       let newInstruction = makePointerToAddress(
         rewritten(s.source), to: RemoteType(t)!, at: s.site)
@@ -331,23 +350,28 @@ extension Module {
     }
   }
 
-  /// Returns the identity of the Val IR function monomorphizing the function referred to by `r`.
-  private mutating func demandMonomorphizedDeclaration(_ r: FunctionReference) -> Function.ID {
-    let result = Function.ID(monomorphized: r.function, for: r.arguments)
+  /// Returns the identity of the Val IR function monomorphizing `f` for given `parameterization`
+  /// in `useScope`.
+  private mutating func demandMonomorphizedDeclaration(
+    of f: Function.ID,
+    for parameterization: GenericArguments,
+    in useScope: AnyScopeID
+  ) -> Function.ID {
+    let result = Function.ID(monomorphized: f, for: parameterization)
     if functions[result] != nil { return result }
 
-    let source = self[r.function]
+    let source = self[f]
 
     let inputs = source.inputs.map { (p) in
-      let t = program.monomorphize(p.type.bareType, for: r.arguments)
+      let t = program.monomorphize(p.type.bareType, for: parameterization)
       return Parameter(decl: p.decl, type: ParameterType(p.type.access, t))
     }
 
-    let output = program.monomorphize(source.output, for: r.arguments)
+    let output = program.monomorphize(source.output, for: parameterization)
 
     let entity = Function(
       isSubscript: source.isSubscript,
-      name: "<\(list: r.arguments.values), \(r.useScope)>(\(source.name))",
+      name: "<\(list: parameterization.values), \(useScope)>(\(source.name))",
       site: source.site,
       linkage: .module,
       parameters: [],

--- a/Sources/IR/Analysis/Module+Depolymorphize.swift
+++ b/Sources/IR/Analysis/Module+Depolymorphize.swift
@@ -23,20 +23,47 @@ extension Module {
   /// counterparts, reading definitions from `ir`.
   private mutating func depolymorphize(_ f: Function.ID, in ir: LoweredProgram) {
     for i in blocks(in: f).map(instructions(in:)).joined() {
-      guard
-        let s = self[i] as? CallInstruction,
-        let callee = s.callee.constant as? FunctionReference,
-        !callee.arguments.isEmpty
-      else { continue }
+      switch self[i] {
+      case let s as CallInstruction:
+        if let new = depolymorphized(s, in: ir) {
+          replace(i, with: new)
+        }
 
-      // TODO: Use existentialization unless the function is inlinable
+      case let s as ProjectInstruction:
+        if let new = depolymorphized(s, in: ir) {
+          replace(i, with: new)
+        }
 
-      let g = monomorphize(callee, in: ir)
-      let r = FunctionReference(to: g, usedIn: callee.useScope, in: self)
-      let newCall = makeCall(
-        applying: .constant(r), to: Array(s.arguments), writingResultTo: s.output, at: s.site)
-      replace(i, with: newCall)
+      default:
+        continue
+      }
     }
+  }
+
+  /// If `s` is a call to a generic function, returns a new instruction applying a depolymorphized
+  /// version of its callee. Otherwise, returns `nil`.
+  private mutating func depolymorphized(
+    _ s: CallInstruction, in ir: LoweredProgram
+  ) -> CallInstruction? {
+    guard
+      let callee = s.callee.constant as? FunctionReference,
+      !callee.arguments.isEmpty
+    else { return nil }
+
+    // TODO: Use existentialization unless the function is inlinable
+
+    let g = monomorphize(callee, in: ir)
+    let r = FunctionReference(to: g, usedIn: callee.useScope, in: self)
+    return makeCall(
+      applying: .constant(r), to: Array(s.arguments), writingResultTo: s.output, at: s.site)
+  }
+
+  /// If `s` is the projection through a generic subscript, returns a new instruction applying a
+  /// depolymorphized version of its callee. Otherwise, returns `nil`.
+  private mutating func depolymorphized(
+    _ s: ProjectInstruction, in ir: LoweredProgram
+  ) -> ProjectInstruction? {
+    return nil
   }
 
   /// Returns a depolymorphized copy of `base` in which parametric parameters have been notionally

--- a/Sources/IR/Emitter.swift
+++ b/Sources/IR/Emitter.swift
@@ -1742,9 +1742,9 @@ public struct Emitter {
     case .direct(let d, _):
       return emitLValue(directReferenceTo: d)
 
-    case .member(let d, _):
+    case .member(let d, let a):
       let r = emitLValue(receiverOf: e)
-      return emitProperty(boundTo: r, declaredBy: d, at: ast[e].site)
+      return emitProperty(boundTo: r, declaredBy: d, parameterizedBy: a, at: ast[e].site)
 
     case .constructor:
       fatalError()
@@ -1790,15 +1790,16 @@ public struct Emitter {
     return emitElementAddr(base, at: [ast[e].index.value], at: ast[e].index.site)
   }
 
-  /// Returns the address of the member declared by `d` and bound to `receiver`, inserting IR
-  /// anchored at `site`.
+  /// Returns the address of the member declared by `d`, parameterized by `a`, and bound to
+  /// `receiver`, inserting IR anchored at `site`.
   private mutating func emitProperty(
-    boundTo receiver: Operand, declaredBy d: AnyDeclID,
+    boundTo receiver: Operand, declaredBy d: AnyDeclID, parameterizedBy a: GenericArguments,
     at site: SourceRange
   ) -> Operand {
     switch d.kind {
     case SubscriptDecl.self:
-      return emitComputedProperty(boundTo: receiver, declaredByBundle: .init(d)!, at: site)
+      return emitComputedProperty(
+        boundTo: receiver, declaredByBundle: .init(d)!, parameterizedBy: a, at: site)
 
     case VarDecl.self:
       let l = AbstractTypeLayout(of: module.type(of: receiver).ast, definedIn: program)
@@ -1810,16 +1811,17 @@ public struct Emitter {
     }
   }
 
-  /// Returns the address of the computed property declared by `d` and bound to `receiver`,
-  /// inserting IR anchored at `site`.
+  /// Returns the projection the property declared by `d`, parameterized by `a`, and bound to
+  /// `receiver`, inserting IR anchored at `site`.
   private mutating func emitComputedProperty(
     boundTo receiver: Operand, declaredByBundle d: SubscriptDecl.ID,
+    parameterizedBy a: GenericArguments,
     at site: SourceRange
   ) -> Operand {
     // TODO: Handle generics
 
     if let i = ast[d].impls.uniqueElement {
-      return emitComputedProperty(boundTo: receiver, declaredBy: i, at: site)
+      return emitComputedProperty(boundTo: receiver, declaredBy: i, parameterizedBy: a, at: site)
     }
 
     let t = SubscriptType(program.relations.canonical(program[d].type))!
@@ -1831,20 +1833,21 @@ public struct Emitter {
     }
 
     return append(
-      module.makeProjectBundle(applying: variants, of: d, typed: t, to: [r], at: site))[0]
+      module.makeProjectBundle(
+        applying: variants, of: d, parameterizedBy: a, typed: t, to: [r], at: site))[0]
   }
 
-  /// Returns the address of the computed property declared by `d` and bound to `receiver`,
-  /// inserting IR anchored at `site`.
+  /// Returns the projection of the property declared by `d`, parameterized by `a`, and bound to
+  /// `receiver`, inserting IR anchored at `site`.
   private mutating func emitComputedProperty(
-    boundTo receiver: Operand, declaredBy d: SubscriptImpl.ID,
+    boundTo receiver: Operand, declaredBy d: SubscriptImpl.ID, parameterizedBy a: GenericArguments,
     at site: SourceRange
   ) -> Operand {
     let t = SubscriptImplType(program[d].type)!
     let o = RemoteType(ast[d].introducer.value, program.relations.canonical(t.output))
     let r = append(module.makeBorrow(o.access, from: receiver, at: site))[0]
     let f = module.demandSubscriptDeclaration(lowering: d)
-    return append(module.makeProject(o, applying: f, to: [r], at: site))[0]
+    return append(module.makeProject(o, applying: f, parameterizedBy: a, to: [r], at: site))[0]
   }
 
   // MARK: Deinitialization

--- a/Sources/IR/Operands/Instruction/ProjectBundleInstruction.swift
+++ b/Sources/IR/Operands/Instruction/ProjectBundleInstruction.swift
@@ -6,6 +6,9 @@ public struct ProjectBundleInstruction: Instruction {
   /// The subscript bundle implementing the projections.
   public let bundle: SubscriptDecl.ID
 
+  /// If `bundle` is generic, the arguments to its generic parameter.
+  public let parameterization: GenericArguments
+
   /// The pure functional type of the callee.
   public let pureCalleeType: LambdaType
 
@@ -24,12 +27,14 @@ public struct ProjectBundleInstruction: Instruction {
   /// Creates an instance with the given properties.
   fileprivate init(
     bundle: SubscriptDecl.ID,
+    parameterization: GenericArguments,
     pureCalleeType: LambdaType,
     variants: [AccessEffect: Function.ID],
     operands: [Operand],
     site: SourceRange
   ) {
     self.bundle = bundle
+    self.parameterization = parameterization
     self.pureCalleeType = pureCalleeType
     self.variants = variants
     self.operands = operands
@@ -77,13 +82,14 @@ extension ProjectBundleInstruction: CustomStringConvertible {
 extension Module {
 
   /// Creates a `project_bundle` anchored at `site` that projects a value by applying one of
-  /// the given `variants` on `arguments`. The variants are defined in `bundle`, which whose
-  /// declaration reference has type `bundleType`.
+  /// the given `variants` on `arguments`. The variants are defined in `bundle`, which is
+  /// parameterized by `parameterization` and has type `bundleType`.
   ///
   /// - Requires: `bundleType` is canonical and `variants` is not empty.
   func makeProjectBundle(
     applying variants: [AccessEffect: Function.ID],
     of bundle: SubscriptDecl.ID,
+    parameterizedBy parameterization: GenericArguments,
     typed bundleType: SubscriptType,
     to arguments: [Operand],
     at site: SourceRange
@@ -91,6 +97,7 @@ extension Module {
     precondition(bundleType[.isCanonical])
     return .init(
       bundle: bundle,
+      parameterization: parameterization,
       pureCalleeType: bundleType.pure,
       variants: variants,
       operands: arguments,

--- a/Sources/IR/Operands/Instruction/ProjectInstruction.swift
+++ b/Sources/IR/Operands/Instruction/ProjectInstruction.swift
@@ -9,6 +9,9 @@ public struct ProjectInstruction: Instruction {
   /// The subscript implementing the projection.
   public let callee: Function.ID
 
+  /// If `callee` is generic, the arguments to its generic parameter.
+  public let parameterization: GenericArguments
+
   /// The arguments of the call.
   ///
   /// Operands to non-`sink` inputs must be the result of a `borrow` instruction requesting the
@@ -22,11 +25,13 @@ public struct ProjectInstruction: Instruction {
   fileprivate init(
     projection: RemoteType,
     callee: Function.ID,
+    parameterization: GenericArguments,
     operands: [Operand],
     site: SourceRange
   ) {
     self.projection = projection
     self.callee = callee
+    self.parameterization = parameterization
     self.operands = operands
     self.site = site
   }
@@ -56,12 +61,21 @@ extension ProjectInstruction: CustomStringConvertible {
 
 extension Module {
 
-  /// Creates a `project` anchored at `site` that projects a value of type `t` by applying `s`
-  /// on `arguments`.
+  /// Creates a `project` anchored at `site` that projects a value of type `t` by applying
+  /// `callee`, which is parameterized by `parameterization`, on `arguments`.
   func makeProject(
-    _ t: RemoteType, applying s: Function.ID, to arguments: [Operand], at site: SourceRange
+    _ t: RemoteType,
+    applying callee: Function.ID,
+    parameterizedBy parameterization: GenericArguments,
+    to arguments: [Operand],
+    at site: SourceRange
   ) -> ProjectInstruction {
-    .init(projection: t, callee: s, operands: arguments, site: site)
+    .init(
+      projection: t,
+      callee: callee,
+      parameterization: parameterization,
+      operands: arguments,
+      site: site)
   }
 
 }

